org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -596,6 +626,72 @@ protected Compiler newCompiler() {
 	return newCompiler;
 }
 
+private CompilerConfiguration prepareCompilerConfiguration(CompilerOptions options) {
+	CompilerConfiguration configuration = new CompilerConfiguration();
+	configuration.setOptions(options);
+	List<String> annotationProcessorPaths = new ArrayList<>();
+	List<String> generatedSourcePaths = new ArrayList<>();
+	boolean isTest = this.compilationGroup == CompilationGroup.TEST;
+	if (this.javaBuilder.participants != null) {
+		for (CompilationParticipant participant : this.javaBuilder.participants) {
+			if (participant.isAnnotationProcessor()) {
+				String[] paths = participant.getAnnotationProcessorPaths(this.javaBuilder.javaProject, isTest);
+				if (paths != null) {
+					annotationProcessorPaths.addAll(Arrays.asList(paths));
+				}
+				String[] generatedSrc = participant.getGeneratedSourcePaths(this.javaBuilder.javaProject, isTest);
+				if (generatedSrc != null) {
+					generatedSourcePaths.addAll(Arrays.asList(generatedSrc));
+				}
+			}
+		}
+	}
+	configuration.setAnnotationProcessorPaths(annotationProcessorPaths);
+	configuration.setGeneratedSourcePaths(generatedSourcePaths);
+	ClasspathLocation[] classpathLocations = this.nameEnvironment.binaryLocations;
+	List<String> classpaths = new ArrayList<>();
+	List<String> modulepaths = new ArrayList<>();
+	for (ClasspathLocation location : classpathLocations) {
+		if (location instanceof ClasspathDirectory cpDirectory) {
+			String filepath = cpDirectory.binaryFolder.getLocation().toFile().getAbsolutePath();
```

Comment:
possible NPE: .getLocation() is documented that it may return null. `if path is null, it means path variable refers to a non-local filesystem`